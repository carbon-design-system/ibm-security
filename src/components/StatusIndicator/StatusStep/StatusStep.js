/**
 * @file Status step.
 * @copyright IBM Security 2019
 */

import Error20 from '@carbon/icons-react/lib/error/20';
import RadioButton20 from '@carbon/icons-react/lib/radio-button/20';

import React, { Component } from 'react';
import classnames from 'classnames';
import { string, oneOf } from 'prop-types';

import { appendComponentNamespace } from '../../../globals/namespace';
import { namespace as indicatorNamespace } from '../StatusIndicator';

import Icon from '../../Icon';
import InlineLoading from '../../InlineLoading';

const namespace = appendComponentNamespace(indicatorNamespace, 'step');

const STATUS = {
  COMPLETE: 'complete',
  CURRENT: 'current',
  FAILED: 'failed',
  DEFAULT: 'incomplete',
};

const getStatusIcon = (status, description) => {
  switch (status) {
    case STATUS.CURRENT:
      return (
        <InlineLoading
          className={`${namespace}-wrapper`}
          description={description}
        />
      );
    case STATUS.COMPLETE:
      return (
        <InlineLoading
          className={`${namespace}-wrapper`}
          description={description}
          status="finished"
        />
      );
    case STATUS.FAILED:
      return (
        <div className={`${namespace}-wrapper`}>
          <div className={`${namespace}__icon-wrapper`}>
            <Icon
              className={classnames(
                `${namespace}__icon ${namespace}--${status}__icon`
              )}
              renderIcon={Error20}
              title={description}
            />
          </div>
          {description && (
            <p className={`${namespace}__label`}>{description}</p>
          )}
        </div>
      );
    default:
      return (
        <div className={`${namespace}-wrapper`}>
          <div className={`${namespace}__icon-wrapper`}>
            <Icon
              className={classnames(
                `${namespace}__icon ${namespace}--${status}__icon`
              )}
              renderIcon={RadioButton20}
              title={description}
            />
          </div>
          <p className={`${namespace}__label`}>{description}</p>
        </div>
      );
  }
};

/**
 * Status Step component.
 */
class StatusStep extends Component {
  static propTypes = {
    /** @type {string} Used to create a custom key for a step. */
    label: string.isRequired,

    /** @type {string} A description for a step. */
    description: string.isRequired,

    /** @type {string} Extra classes to add. */
    className: string,

    /** @type {enum} A status for a step. */
    status: oneOf([
      STATUS.COMPLETE,
      STATUS.CURRENT,
      STATUS.DEFAULT,
      STATUS.FAILED,
    ]),

    /** @type {string} An error message to explain why the step failed. */
    errorMsg: string,
  };

  static defaultProps = {
    className: '',
    status: STATUS.DEFAULT,
    errorMsg: null,
  };

  state = {
    status: this.props.status,
  };

  static getDerivedStateFromProps(nextProps, prevProps) {
    if (nextProps.status && prevProps.status !== nextProps.status)
      return {
        status: nextProps.status,
      };
    return null;
  }

  render() {
    const { label, description, className, errorMsg } = this.props;

    const stepClasses = classnames(
      namespace,
      `${namespace}--${this.state.status}`,
      className
    );

    return (
      <li className={stepClasses} key={`${namespace}--${label}`}>
        {getStatusIcon(this.state.status, description)}
        {this.state.status === STATUS.FAILED && errorMsg && (
          <p className={`${namespace}__error-message`}>{errorMsg}</p>
        )}
      </li>
    );
  }
}

export default StatusStep;

export { namespace, STATUS };
