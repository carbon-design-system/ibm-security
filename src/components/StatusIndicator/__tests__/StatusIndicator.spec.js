/**
 * @file Status step tests.
 * @copyright IBM Security 2019 - 2021
 */

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { StatusIndicator, StatusStep } from '../../..';

import { STATUS } from '../StatusStep/StatusStep';

describe('StatusIndicator', () => {
  test('should have no Axe or DAP violations', async () => {
    const { container } = render(
      <StatusIndicator
        title="test title"
        retry={{
          label: 'test retry label',
          action: () => {},
        }}
      >
        <StatusStep label="test-label-1" description="test description 1" />
        <StatusStep
          status={STATUS.current}
          label="test-label-2"
          description="test description 2"
        />
        <StatusStep
          status={STATUS.complete}
          label="test-label-3"
          description="test description 3"
        />
        <StatusStep
          status={STATUS.error}
          errorMsg="test error"
          label="test-label-4"
          description="test description 4"
        />
      </StatusIndicator>
    );
    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations(`StatusIndicator`);
  });

  test('should add a custom class', () => {
    render(<StatusIndicator className="custom-class" />);
    expect(document.querySelector('.custom-class')).toBeVisible();
  });

  test('should add a title', () => {
    const { queryByText } = render(<StatusIndicator title="test title" />);
    expect(queryByText(/test title/i)).toBeVisible();
  });

  test('should add children', () => {
    const { getByText } = render(
      <StatusIndicator>
        <span>test text</span>
      </StatusIndicator>
    );
    expect(getByText(/test text/i)).toBeVisible();
  });

  test('should add a `retry` object with a `label`', () => {
    const { queryByText } = render(
      <StatusIndicator
        retry={{
          label: 'test retry label',
          action: () => {},
        }}
      />
    );
    expect(queryByText(/test retry label/i)).toBeVisible();
  });

  test('should invoke onClick mock when `retry.action` is clicked', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <StatusIndicator
        retry={{
          label: 'test retry label',
          action: onClickMock,
        }}
      />
    );
    userEvent.click(getByText(/test retry label/i));
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });
});
