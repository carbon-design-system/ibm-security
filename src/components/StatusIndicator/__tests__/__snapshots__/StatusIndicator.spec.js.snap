// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusIndicator renders 1`] = `
<div
  className="security--status-indicator"
>
  <h1
    className="security--status-indicator__title"
  >
    Status indicator title
  </h1>
  <ul
    className="security--status-indicator__list"
  >
    <StatusStep
      className=""
      description="Status step 1"
      errorMsg={null}
      key="0"
      label="Status step 1"
      status="complete"
    />
    <StatusStep
      className=""
      description="Status step 2"
      errorMsg={null}
      key="1"
      label="Status step 2"
      status="current"
    />
    <StatusStep
      className=""
      description="Status step 3"
      errorMsg={null}
      key="2"
      label="Status step 3"
      status="failed"
    />
    <StatusStep
      className=""
      description="Status step 4"
      errorMsg={null}
      key="3"
      label="Status step 4"
      status="incomplete"
    />
    <Button
      className="security--status-indicator__button--retry"
      disabled={false}
      kind="ghost"
      loading={false}
      renderIcon={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      tabIndex={0}
      type="button"
    >
      Retry
    </Button>
  </ul>
</div>
`;
